Setting the initial pgsql superuser password

robdaemon opened the pull request:
* Adds initial-su-username and initial-su-password-secret-file parameters. If these are set, stolon will set the proper password for the initial superuser when postgresql is first initialized.
* Adds an example secret and usage to the Kubernetes examples.
* Adds support for specifying a namespace to sentinel. This allows stolon to be ran outside of the "default" namespace.

sgotti commented on cmd/keeper/keeper.go:
```diff
@@ -93,6 +95,8 @@ func init() {
 	cmdKeeper.PersistentFlags().StringVar(&cfg.pgSUUsername, "pg-su-username", "", "postgres superuser user name (required by pg_rewind)")
 	cmdKeeper.PersistentFlags().StringVar(&cfg.pgSUPassword, "pg-su-password", "", "postgres superuser password (required by pg_rewind)")
+	cmdKeeper.PersistentFlags().StringVar(&cfg.initialSUUsername, "initial-su-username", "", "postgres initial superuser username")
+	cmdKeeper.PersistentFlags().StringVar(&cfg.initialSUPasswordFile, "initial-su-password-secret-file", "", "postgres initial superuser password secret file")
```
I also wanted to read it from a file when adding `pgSUPassword` for pg rewind instead of relying of an argument or env variable. I'll create a future PR the make it consistent with your proposal.

sgotti replied:
To make it consistent I'd like to leave the pg prefix. But the options are becoming quite long :stuck_out_tongue_closed_eyes: 

what about:

`initial-pg-su-username`
`initial-pg-su-password-file`
?

sgotti commented on cmd/keeper/keeper.go:
```diff
@@ -113,9 +131,6 @@ func (p *PostgresKeeper) getSUConnParams(keeperState *cluster.KeeperState, setPa
 	}
-	if setPassword {
-		cp["password"] = p.cfg.pgSUPassword
-	}
```
Not sure why I addedt the setPassword bool but it's actually not needed... thanks.

sgotti commented on cmd/keeper/keeper.go:
```diff
@@ -103,7 +107,21 @@ var defaultPGParameters = pg.Parameters{
+	var pw string
+
+	if p.cfg.initialSUPasswordFile != "" {
+		pwBytes, err := ioutil.ReadFile(p.cfg.initialSUPasswordFile)
```
should we check file permissions to be <= than 600. Perhaps that's a lesser problem inside a pod but on a standard server it'll be good to make them readable only by the stolon's owner.

But looks like that kubernetes actually doesn't implement setting file permission so actually it should ignore the check on kubernetes. Perhaps just add a TODO at the moment...
Haven't tried so I'm not sure it will be possible to them with that user / permissions using kubernetes secrets?

sgotti commented on cmd/keeper/keeper.go:
```diff
@@ -103,7 +107,21 @@ var defaultPGParameters = pg.Parameters{
+	if p.cfg.initialSUPasswordFile != "" {
+		pwBytes, err := ioutil.ReadFile(p.cfg.initialSUPasswordFile)
+		if err != nil {
+			log.Errorf("Unable to read password from file %s, error: %s", p.cfg.initialSUPasswordFile, err.Error())
```
`log.Errorf("Unable to read password from file %s, error: %v", p.cfg.initialSUPasswordFile, err)` to make it consistent with other errors logging.

sgotti commented:
@robdaemon thanks a lot for this patch, that's really cool. I put some comments inline.

robdaemon commented:
Give me an hour or so, I'll push up a commit that addresses your comments!

robdaemon commented:
@sgotti okay, addressed your feedback. I added a TODO about enforcing the file permissions, and a log warning if they're too open. I chose a warning on the logging because it's a recoverable condition.

sgotti commented:
@robdaemon Great! Thanks a lot! I'm going to merge it.

sgotti commented:
@robdaemon I forgot to ask if you can please squash the commits in a single one.

robdaemon commented:
done! I usually hold off on squashing until it meets upstream approval :)

sgotti commented:
> done! I usually hold off on squashing until it meets upstream approval :)

You're right, I was just too eager to merge it.

Thanks again!
